doc/arch/adr-002-decide-how-to-display-more-facilities.md

```diff
@@ -0,0 +1,232 @@
+
+An `ST_AsMVT`-based approach to generate vector tiles dynamically seemed to be
+promising. While the vector tiles working group's report did note some
+uncertainty around how performant it would be to generate tiles in PostGIS,
+OAR's traffic is such that it may not encounter performance problems which
+could emerge for a higher traffic site.
+
+We considered two ways to generate tiles using `ST_AsMVT`:
+
+- using a dedicated vector tile server like [Martin][martin], [t-rex][trex], or
+[tegola][tegola]
+- adding a vector tiles endpoint to the existing Django web app
+
+### Using Martin (or t-rex or Tegola) as a Vector Tile Server
+
+Martin, t-rex, and Tegola are open-source vector tile servers which can connect
+directly to PostGIS and render vector tiles. Judging by their documentation,
+each of them appear to be fairly straightforward to configure and operate. Each
+has a slightly different API.
+
+We considered Martin most seriously as an option in part because it had good
+documentation around how to write PL/pgSQL functions for requesting tiles with
+data filtered by a set of query parameters. Here's [Martin's function sources example][martin-function-sources]:
+
+```plpgsql
+CREATE OR REPLACE FUNCTION public.function_source(z integer, x integer, y integer, query_params json) RETURNS BYTEA AS $$
+DECLARE
+  bounds GEOMETRY(POLYGON, 3857) := TileBBox(z, x, y, 3857);
+  mvt BYTEA;
+BEGIN
+  SELECT INTO mvt ST_AsMVT(tile, 'public.function_source', 4096, 'geom') FROM (
+    SELECT
+      ST_AsMVTGeom(geom, bounds, 4096, 64, true) AS geom
+    FROM public.table_source
+    WHERE geom && bounds
+  ) as tile WHERE geom IS NOT NULL;
+
+  RETURN mvt;
+END
+$$ LANGUAGE plpgsql IMMUTABLE STRICT PARALLEL SAFE;
+```
+
+#### Pros
+
+##### Configuration
+
+Martin appears fairly straightforward to configure and its documentation
+encompassed most of what we'd want to do.
+
+##### Performance (?)
+
+Martin touts being "suitable for large databases" which indicates that it
+might obviate some of the performance concerns around using `ST_AsMVT`.
+
+#### Cons
+
+##### PL/pgSQL Function Sources
+
+Since Martin uses PL/pgSQL functions for its filtering, we would have to
+rewrite some facility filtering logic that currently exists in Django querysets
+in the web app to work in PL/pgSQL. Moreover, each time we added a new filter
+or search option to the web application, we'd have to write a version of the
+same query in PL/pgSQL for the tiler.
+
+##### Security
+
+Martin's `query_params` appear to be passed in to the database as strings, which
+opens a security hole. While we could create a PostGIS role or user with a
+limited, readonly set of permissions to use solely for the Martin instance,
+doing so requires taking on some additional risk and complexity.
+
+##### Unfamiliarity
+
+We don't have any experience running Martin in production. We've also got
+limited experience using Rust. Together this means a Martin-based tile server
```

Comment:
A sentence here or at the top of the section that Martin is written in Rust would help clarify

Reply:
Oh, great point -- I'll add a phrase somewhere to note that Martin is written in Rust.